tests/integration/test_upgrade_charm.py

```diff
@@ -28,12 +52,77 @@ async def test_deploy_from_edge_and_upgrade_from_local_path(ops_test, prometheus
     """
     async with IPAddressWorkaround(ops_test):
         logger.debug("deploy charm from charmhub")
-        await ops_test.model.deploy(f"ch:{app_name}", application_name=app_name, channel="edge")
-        await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=TIMEOUT)
+        await ops_test.model.deploy(
+            f"ch:{PROMETHEUS_APP_NAME}", application_name=PROMETHEUS_APP_NAME, channel="edge"
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[PROMETHEUS_APP_NAME], status="active", timeout=TIMEOUT
+        )
 
         logger.debug("upgrade deployed charm with local charm %s", prometheus_charm)
-        await ops_test.model.applications[app_name].refresh(
-            path=prometheus_charm, resources=resources
+        await ops_test.model.applications[PROMETHEUS_APP_NAME].refresh(
+            path=prometheus_charm, resources=prometheus_resources
         )
-        await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=TIMEOUT)
-        await check_prometheus_is_ready(ops_test, app_name, 0)
+        await ops_test.model.wait_for_idle(
+            apps=[PROMETHEUS_APP_NAME], status="active", timeout=TIMEOUT
+        )
+        await check_prometheus_is_ready(ops_test, PROMETHEUS_APP_NAME, 0)
+
+
+async def test_metrics_endpoint_provider_can_be_upgraded(ops_test, prometheus_tester_charm):
+    """Ensure scrape alert rules can be updated.
+
+    This test upgrades the metrics provider charm and checks that
+    updates to alert rules are propagated correctly.
+    """
+    await ops_test.model.deploy(
+        prometheus_tester_charm, resources=tester_resources, application_name=TESTER_APP_NAME
+    )
```

Comment:
I think the tester should be deployed and related to prom _before_ prom gets refreshed from local path (line 63).

Reply:
The purpose of the new test is check that MetricsEndpointProvider upgrade path works since the bug described in the ticket was that new alert rules were not being set after upgrade. If modifications are required to the previously existing test for MetricsEndpointConsumer then we can create a separate ticket and modify as part of that ticket.

Reply:
I was not aware of that. Can you please add the bug description or crossref to the ticket?

Reply:
In meeting with @mmanciop the discussion concluded it would be useful to have a integration test to check that alert rules are updated on charm upgrade and that the alert rules check was sufficient (i.e without scrape job check) . The alert rules are updated by `MetricsEndpointProvider` **not** `MetricsEndpointConsumer`. The only issue on github was #188, which I have now clarified and closed.

Reply:
Thanks for your time @sed-i . As discussed I have
- Created #234 to enhance the existing `MetricsEnpointConsumer` test in the next commit.
- On refactoring the `MetricsEndpointConsumer` test will be renamed to emphasise it is testing the Consumer charm.
- Also as discussed I will retain both tests in the same module `test_upgrade_charm.py` and document that this module test upgrade of both Provider and Consumer charm.